[GlobalISel][AArch64] Add G_FPTOSI_SAT/G_FPTOUI_SAT

davemgreen opened the pull request:
This is an implementation of the saturating fp to int conversions for GlobalISel. On AArch64 the converstion instrctions work this way, producing saturating results.  LegalizerHelper::lowerFPTOINT_SAT is ported from SDAG.

AArch64 has a lot of existing tests for fptosi_sat, covering a wide range of types. I have tried to make most of them work all at once, but a few fall back due to other missing features such as f128 handling for min/max.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7012,6 +7063,102 @@ LegalizerHelper::LegalizeResult LegalizerHelper::lowerFPTOSI(MachineInstr &MI) {
+    // Clamp Src by MinFloat from below. If Src is NaN the result is MinFloat.
+    auto Max = MIRBuilder.buildFMaxNum(
+        SrcTy, Src, MIRBuilder.buildFConstant(SrcTy, MinFloat));
+    // Clamp by MaxFloat from above. NaN cannot occur.
```
Then add the nnan flag?

tschuett commented:
Could you link/fixes https://github.com/llvm/llvm-project/issues/94694 and maybe take the test case?

davemgreen commented:
> Could you link/fixes https://github.com/llvm/llvm-project/issues/94694 and maybe take the test case?

I believe that whilst #94694 is currently related to fptosi.sat, the same issue with infinite combines would come up from any other generic intrinsic. Prior to this patch GISel didn't know anything special about fptosi.sat, so it was just using a G_INTRINSIC and I believe the same problem would still exist for anything that wasn't now transformed into a G_FPTOSI_SAT.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -2697,6 +2700,47 @@ LegalizerHelper::widenScalar(MachineInstr &MI, unsigned TypeIdx, LLT WideTy) {
+            WideTy, APInt::getSignedMaxValue(ShortBits).sext(WideBits));
+        auto MinVal = MIRBuilder.buildConstant(
+            WideTy, APInt::getSignedMinValue(ShortBits).sext(WideBits));
+        Register MidReg =
+            MIRBuilder.buildSMin(WideTy, ExtReg, MaxVal).getReg(0);
```
auto and remove the getReg(0)

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7012,6 +7063,103 @@ LegalizerHelper::LegalizeResult LegalizerHelper::lowerFPTOSI(MachineInstr &MI) {
+  // If the integer bounds are exactly representable as floats and min/max are
+  // legal, emit a min+max+fptoi sequence. Otherwise we have to use a sequence
+  // of comparisons and selects.
+  bool MinMaxLegal = LI.isLegal({TargetOpcode::G_FMINNUM, SrcTy}) &&
+                     LI.isLegal({TargetOpcode::G_FMAXNUM, SrcTy});
```
Exact legality isn't exactly the right check. Ideally we would figure out if the operation would ultimately be lowered to a fmin/fmax. 

Looking at this again, below you are using fcmp / select. We should probably just pick one or the other as the canonical form, and let the nnan help optimizations convert into the appropriate flavor of min/max

arsenm replied:
Yes, I don't think this should be conditionalized. Just pick one as the canonical no-nan min/max case. If we need the conditions, it should be done as a separate change.

We do have a possible issue that we don't have an optimizing combiner mid-legalize, so the target won't get a chance to morph whatever the canonical nnan fmin/fmax is into whichever form it prefers

arsenm replied:
ping here, the legality shouldn't matter. Do the cmp + select with nnan and let the combiners figure it out?

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:
```diff
@@ -2130,6 +2130,12 @@ bool AArch64InstructionSelector::preISelLower(MachineInstr &I) {
     return false;
   }
+  case TargetOpcode::G_FPTOSI_SAT:
+    I.setDesc(TII.get(TargetOpcode::G_FPTOSI));
```
I would hope this would have been a trivial selection pattern?

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -708,6 +708,55 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .libcallFor(
           {{s32, s128}, {s64, s128}, {s128, s128}, {s128, s32}, {s128, s64}});
 
+  getActionDefinitionsBuilder({G_FPTOSI_SAT, G_FPTOUI_SAT})
```
This is one complicated rule set!

davemgreen commented:
Rebase and ping, thanks.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7285,6 +7335,106 @@ LegalizerHelper::LegalizeResult LegalizerHelper::lowerFPTOSI(MachineInstr &MI) {
+  // is already zero.
+  if (!IsSigned) {
+    MIRBuilder.buildSelect(Dst, OGT, MIRBuilder.buildConstant(DstTy, MaxInt),
+                           Max, MachineInstr::FmNoNans);
```
Should the fcmp also get the flag?

davemgreen replied:
IIRC the FCmp is based on the original value, so could still be Nan. This select appears to be integer though, so presumably shouldn't have FMF's.

davemgreen committed:
[GlobalISel][AArch64] Add G_FPTOSI_SAT/G_FPTOUI_SAT

This is an implementation of the saturating fp to int conversions for
GlobalISel. On AArch64 the converstion instrctions work this way, producing
saturating results.  LegalizerHelper::lowerFPTOINT_SAT is ported from SDAG.

AArch64 has a lot of existing tests for fptosi_sat, covering a wide range of
types. I have tried to make most of them work all at once, but a few fall back
due to other missing features such as f128 handling for min/max.

jayfoad commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -569,6 +569,8 @@ def fp_to_sint : SDNode<"ISD::FP_TO_SINT" , SDTFPToIntOp>;
 def fp_to_uint : SDNode<"ISD::FP_TO_UINT" , SDTFPToIntOp>;
 def fp_to_sint_sat : SDNode<"ISD::FP_TO_SINT_SAT" , SDTFPToIntSatOp>;
 def fp_to_uint_sat : SDNode<"ISD::FP_TO_UINT_SAT" , SDTFPToIntSatOp>;
+def fp_to_sint_sat_gi : SDNode<"ISD::FP_TO_SINT_SAT" , SDTFPToIntOp>;
+def fp_to_uint_sat_gi : SDNode<"ISD::FP_TO_UINT_SAT" , SDTFPToIntOp>;
```
@davemgreen could we somehow avoid introducing these new node types, and just use the same ones as SelectionDAG? I realize the SelectionDAG ones are more powerful in that they allow you to specify a saturate-to type that is smaller than the result type.

Currently targets have to write more-or-less the same selections patterns twice, once with `fp_to_sint_sat` and once with `fp_to_sint_sat_gi`.

jayfoad replied:
Sorry I never followed up on this. It seems @IgWod has picked it up now with #202908.
